torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -0,0 +1,101 @@
+    dtype: torch.dtype,
+) -> Optional[AllGatherResult]:
+    world_size, rank = (group.size(), group.rank())
+    # - Copy in
+    with torch.cuda.stream(all_gather_copy_in_stream):
+        param_all_gather_inputs = [
+            fsdp_param.all_gather_input for fsdp_param in fsdp_params
+        ]
+        inp_split_sizes = [inp.numel() for inp in param_all_gather_inputs]
+        all_gather_input_numel = sum(inp_split_sizes)
+        all_gather_output = torch.empty(
+            (all_gather_input_numel * world_size,), dtype=dtype, device=device
+        )
+        all_gather_input = all_gather_output.narrow(
+            0, all_gather_input_numel * rank, all_gather_input_numel
+        )
+        foreach_copy_dsts = torch.split(all_gather_input, inp_split_sizes)
+        torch._foreach_copy_(foreach_copy_dsts, param_all_gather_inputs)
+        del param_all_gather_inputs
+        all_gather_copy_in_event = torch.cuda.Event()
+        all_gather_copy_in_event.record()
+    all_gather_stream.wait_event(all_gather_copy_in_event)
+    with torch.cuda.stream(all_gather_stream):
+        # - All-gather
+        all_gather_work = dist.all_gather_into_tensor(
+            output_tensor=all_gather_output,
+            input_tensor=all_gather_input,
+            group=group,
+            async_op=async_op,
+        )
+        all_gather_event = torch.cuda.Event()
+        all_gather_event.record()
+        return AllGatherResult(
+            all_gather_output, all_gather_event, all_gather_work, inp_split_sizes
+        )
+
+
+def foreach_all_gather_copy_out(
+    all_gather_result: AllGatherResult,
+    fsdp_params: List[FSDPParam],
+    group: dist.ProcessGroup,
+) -> None:
+    all_gather_output, _, _, all_gather_input_numels = all_gather_result
+    if (event := all_gather_result.all_gather_event) is not None:  # sync op
+        torch.cuda.current_stream().wait_event(event)
+    if (work := all_gather_result.all_gather_work) is not None:  # async op
+        work.wait()
+    world_size = group.size()
+    dtype, device = all_gather_output.dtype, all_gather_output.device
+    for all_gather_input_numel, fsdp_param in zip(all_gather_input_numels, fsdp_params):
+        fsdp_param.init_all_gather_output(
+            all_gather_input_numel, world_size, dtype, device
+        )  # no-op after 1st call
+        fsdp_param.alloc_all_gather_output()
+    all_gather_output = all_gather_output.view(world_size, -1)
+    out = [
+        fsdp_param.all_gather_output.view(world_size, -1) for fsdp_param in fsdp_params
+    ]
+    # TODO: Use `torch.split_with_sizes_copy` fast path once it lands.
+    splits = torch.split(all_gather_output, all_gather_input_numels, dim=1)
+    with _unsafe_preserve_version_counters(out):
+        torch._foreach_copy_(out, splits)  # one `copy_` per parameter
+
+
+class _unsafe_preserve_version_counters(_DecoratorContextManager):
+    # Same as `_unsafe_preserve_version_counter` but only entering/exiting the
```

Comment:
``_unsafe_preserve_version_counter`` is still mysterious to me. maybe move the PR description here?

``otherwise autograd will complain of a version mismatch in backward due to writing to the leaf tensors``

Reply:
commenting about the runtime error sounds even more clear. thanks